File: src/types/sla.ts
import { Priority } from './request';

export type SLATier = 'PLATINUM' | 'GOLD' | 'SILVER' | 'STANDARD';

export interface PriorityTarget {
  priority: Priority;
  responseTimeMinutes: number;
  resolutionTimeHours: number;
  escalationWarningMinutes: number;
}

export interface SLAPolicy {
  id: string;
  name: string;
  tier: SLATier;
  description: string;
  businessHours: '24x7' | '9-to-5' | 'Custom';
  targets: PriorityTarget[];
  breachPenaltyEnabled: boolean;
  breachPenaltyPerMinuteUsd?: number;
  departmentId?: string;
  isDefault?: boolean;
}

export type SLAStatus = 'MET' | 'ON_TRACK' | 'WARNING' | 'AT_RISK' | 'BREACHED';

export interface SLAProgress {
  status: SLAStatus;
  percentageElapsed: number; // 0 to 100+
  minutesRemaining: number;
  formattedRemaining: string;
  isBreached: boolean;
  totalDurationMinutes: number;
  elapsedMinutes: number;
}

export interface DepartmentPerformance {
  department: string;
  totalRequests: number;
  metSlaCount: number;
  breachedCount: number;
  atRiskCount: number;
  complianceRate: number; // 0 to 100
  avgResolutionHours: number;
  healthStatus: 'HEALTHY' | 'WARNING' | 'CRITICAL';
}

File: backend/app/schemas/sla.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field

class SLATargetSchema(BaseModel):
    priority: str
    targetResponseMinutes: int
    targetResolutionHours: int
    escalationThresholdMinutes: int

class SLAPolicySchema(BaseModel):
    id: str
    name: str
    tier: str  # PLATINUM, GOLD, SILVER, STANDARD
    description: Optional[str] = None
    businessHours: str = "24x7"
    breachPenaltyEnabled: bool = True
    breachPenaltyPerMinuteUsd: float = 50.0
    isDefault: bool = False
    targets: List[Dict[str, Any]] = Field(default_factory=list)

    class Config:
        from_attributes = True

class SLAPolicyUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    businessHours: Optional[str] = None
    breachPenaltyEnabled: Optional[bool] = None
    breachPenaltyPerMinuteUsd: Optional[float] = None
    isDefault: Optional[bool] = None
    targets: Optional[List[Dict[str, Any]]] = None
